File: sdks/dotnet/src/Mobiscroll.Connect/Exceptions/CalendarPermissionException.cs
using System.Collections.Generic;
using Mobiscroll.Connect.Models;

namespace Mobiscroll.Connect.Exceptions;

/// <summary>
/// Thrown when no connected account has the calendar access the request needs.
/// </summary>
/// <remarks>
/// The user completed sign-in but did not grant the calendar permission — Google's consent
/// screen presents it as a separate checkbox. This cannot be repaired server-side, because
/// providers only issue permissions at consent time: the accounts in <see cref="Accounts"/>
/// have to run the connect flow again and allow access.
///
/// Derives from <see cref="AuthenticationException"/>, so existing handlers keep working.
/// </remarks>
public sealed class CalendarPermissionException : AuthenticationException
{
    public CalendarPermissionException(string message, IReadOnlyList<BlockedAccount>? accounts = null)
        : base(message, "CALENDAR_PERMISSION_REQUIRED")
    {
        Accounts = accounts ?? new List<BlockedAccount>();
    }

    /// <summary>Connected accounts that withheld calendar access and must reconnect.</summary>
    public IReadOnlyList<BlockedAccount> Accounts { get; }
}

File: sdks/java/connect-sdk/src/main/java/com/mobiscroll/connect/exceptions/CalendarPermissionException.java
package com.mobiscroll.connect.exceptions;

import java.util.Collections;
import java.util.List;

import com.mobiscroll.connect.models.BlockedAccount;

/**
 * Thrown when no connected account has the calendar access the request needs.
 *
 * <p>The user completed sign-in but did not grant the calendar permission &mdash; Google's
 * consent screen presents it as a separate checkbox. This cannot be repaired server-side,
 * because providers only issue permissions at consent time: the accounts returned by
 * {@link #getAccounts()} have to run the connect flow again and allow access.
 *
 * <p>Extends {@link AuthenticationException}, so existing handlers keep working.
 */
public class CalendarPermissionException extends AuthenticationException {

    private static final long serialVersionUID = 1L;

    private final transient List<BlockedAccount> accounts;

    public CalendarPermissionException(String message, List<BlockedAccount> accounts) {
        super(message);
        this.accounts = accounts == null ? Collections.emptyList() : Collections.unmodifiableList(accounts);
    }

    /** Connected accounts that withheld calendar access and must reconnect. */
    public List<BlockedAccount> getAccounts() {
        return accounts;
    }
}
